src/coreclr/vm/riscv64/stubs.cpp

```diff
@@ -1525,68 +1534,461 @@ static void LoadRegPair(BYTE* p, int reg1, int reg2, UINT32 offset)
+    argsWriterHolder.GetRW()->signature = pLookup->signature;
+    argsWriterHolder.GetRW()->module = (CORINFO_MODULE_HANDLE)pModule;
+
+    WORD slotOffset = (WORD)(dictionaryIndexAndSlot & 0xFFFF) * sizeof(Dictionary*);
+
+    // It's available only via the run-time helper function
+    if (pLookup->indirections == CORINFO_USEHELPER)
+    {
+        BEGIN_DYNAMIC_HELPER_EMIT(32);
+
+        // a0 already contains generic context parameter
+        // reuse EmitHelperWithArg for below two operations
+        // a1 <- pArgs
+        // branch to helperAddress
+        EmitHelperWithArg(p, rxOffset, pAllocator, (TADDR)pArgs, helperAddress);
+
+        END_DYNAMIC_HELPER_EMIT();
+    }
+    else
+    {
+        int codeSize = 0;
+        int indirectionsDataSize = 0;
+        if (pLookup->testForNull || pLookup->sizeOffset != CORINFO_NO_SIZE_CHECK)
+        {
+            codeSize += 4; //mv t2, 0
+        }
+
+        for (WORD i = 0; i < pLookup->indirections; i++) {
+            _ASSERTE(pLookup->offsets[i] >= 0);
+            if (i == pLookup->indirections - 1 && pLookup->sizeOffset != CORINFO_NO_SIZE_CHECK)
+            {
+                codeSize += (pLookup->sizeOffset > 2047 ? 28 : 16);
+                indirectionsDataSize += (pLookup->sizeOffset > 2047 ? 4 : 0);
+            }
+
+            codeSize += (pLookup->offsets[i] > 2047 ? 16 : 4); // if( > 2047) (8 bytes) else 4 bytes for instructions.
+            indirectionsDataSize += (pLookup->offsets[i] > 2047 ? 4 : 0); // 4 bytes for storing indirection offset values
+        }
+
+        codeSize += indirectionsDataSize ? 4 : 0; // auipc
+
+        if (pLookup->testForNull)
+        {
+            codeSize += 12; // ori-beq-jalr
+
+            //padding for 8-byte align (required by EmitHelperWithArg)
+            if (codeSize & 0x7)
+                codeSize += 4;
+
+            codeSize += 32; // size of EmitHelperWithArg
+        }
+        else
+        {
+            codeSize += 4; /* jalr */
+        }
+
+        // the offset value of data_label.
+        uint dataOffset = codeSize;
+
+        codeSize += indirectionsDataSize;
+
+        BEGIN_DYNAMIC_HELPER_EMIT(codeSize);
+
+        BYTE * old_p = p;
+
+        if (indirectionsDataSize)
+        {
+            // auipc  t4, 0
+            *(DWORD*)p = 0x00000e97;
+            p += 4;
+        }
+
+        if (pLookup->testForNull || pLookup->sizeOffset != CORINFO_NO_SIZE_CHECK)
+        {
+            // mv  t2, a0
+            *(DWORD*)p = 0x00050393;
+            p += 4;
+        }
+
+        BYTE* pBLECall = NULL;
+
+        for (WORD i = 0; i < pLookup->indirections; i++)
+        {
+            if (i == pLookup->indirections - 1 && pLookup->sizeOffset != CORINFO_NO_SIZE_CHECK)
+            {
+                _ASSERTE(pLookup->testForNull && i > 0);
+
+                if (pLookup->sizeOffset > 2047)
+                {
+                    // auipc  t4, 0
+                    *(DWORD*)p = 0x00000297; p += 4;
```

Comment:
Isn't it better to use the previous t4 which is set when indirectionsDataSize is over zero? You can remove  many auipc instructions, simplify complex dataOffset calculation and handle indirectionsDataSize as boolean variable. (However, You need to rename t4 which is used for lui.) If it is impossible, please remove previous `auipc t4, 0`.

Reply:
I agree. This will remove both `auipc` instructions in for loop, and `dataOffset` will be relative not to current instruction, but to the beginning of whole code block. Also, it would be good to add assert that `dataOffset` relative to beginning of code block is <= 2047.

Reply:
Comment with more details about this logic in https://github.com/dotnet/runtime/pull/94766#discussion_r1394005571.

Reply:
Done